near-contract-standards/src/non_fungible_token/event.rs

```diff
@@ -0,0 +1,174 @@
+    #[serde(flatten)]
+    pub event_kind: Nep171EventKind,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(tag = "event", content = "data")]
+#[serde(rename_all = "snake_case")]
+#[allow(clippy::enum_variant_names)]
+pub enum Nep171EventKind {
+    NftMint(Vec<NftMintData>),
+    NftTransfer(Vec<NftTransferData>),
+    NftBurn(Vec<NftBurnData>),
+}
+
+#[skip_serializing_none]
+#[derive(Serialize, Deserialize, Debug)]
+pub struct NftMintData {
+    pub owner_id: String,
+    pub token_ids: Vec<String>,
+    pub memo: Option<String>,
+}
+
+#[skip_serializing_none]
+#[derive(Serialize, Deserialize, Debug)]
+pub struct NftTransferData {
+    pub authorized_id: Option<String>,
+    pub old_owner_id: String,
+    pub new_owner_id: String,
+    pub token_ids: Vec<String>,
+    pub memo: Option<String>,
+}
+
+#[skip_serializing_none]
+#[derive(Serialize, Deserialize, Debug)]
+pub struct NftBurnData {
+    pub authorized_id: Option<String>,
+    pub owner_id: String,
+    pub token_ids: Vec<String>,
+    pub memo: Option<String>,
+}
+
+impl Display for NearEvent {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(&format!("EVENT_JSON:{}", self.to_json_string()))
+    }
+}
+
+use Nep171EventKind::*;
+
+impl NearEvent {
+    pub fn new_171(version: String, event_kind: Nep171EventKind) -> Self {
+        NearEvent::Nep171(Nep171Event { version, event_kind })
+    }
+
+    pub fn new_171_v1(event_kind: Nep171EventKind) -> Self {
+        NearEvent::new_171("1.0.0".to_string(), event_kind)
+    }
+
+    pub fn nft_burn(data: Vec<NftBurnData>) -> Self {
+        NearEvent::new_171_v1(NftBurn(data))
+    }
+    pub fn nft_transfer(data: Vec<NftTransferData>) -> Self {
+        NearEvent::new_171_v1(NftTransfer(data))
+    }
+
+    pub fn nft_mint(data: Vec<NftMintData>) -> Self {
+        NearEvent::new_171_v1(NftMint(data))
+    }
+
+    pub(crate) fn to_json_string(&self) -> String {
+        serde_json::to_string(self).unwrap()
+    }
+
+    pub fn log(&self) {
+        near_sdk::env::log_str(&self.to_string());
+    }
+
+    pub fn log_nft_mint(owner_id: String, token_ids: Vec<String>, memo: Option<String>) {
+        NearEvent::nft_mint(vec![NftMintData { owner_id, token_ids, memo }]).log();
+    }
+    pub fn log_nft_transfer(
+        old_owner_id: String,
+        new_owner_id: String,
+        token_ids: Vec<String>,
+        memo: Option<String>,
+        authorized_id: Option<String>,
+    ) {
+        NearEvent::nft_transfer(vec![NftTransferData {
+            authorized_id,
+            old_owner_id,
+            new_owner_id,
+            token_ids,
+            memo,
+        }])
+        .log();
+    }
+
+    pub fn log_nft_burn(
+        owner_id: String,
+        authorized_id: Option<String>,
+        token_ids: Vec<String>,
+        memo: Option<String>,
+    ) {
+        NearEvent::nft_burn(vec![NftBurnData { owner_id, authorized_id, token_ids, memo }]).log();
+    }
```

Comment:
This API has a few problems:

1. The usage of these methods will be error-prone to use and hard to reason about:
    ```rust
    NearEvent::log_nft_transfer("frol.near", "willem.near", vec!["dacha".to_string()], Some("qq".to_string()), Some("alex.near"))
    NearEvent::log_nft_burn("willem.near", Some("alex.near"), vec!["dacha".to_string()], Some("qq".to_string()))
    ```
  
    (You already mixed up the order of `authorized_id` in transfer and burn methods)
2. These methods promote non-batched events and thus we promote using more gas on events than necessary

Reply:
1. Yeah oops, this was from just following the ordering from the struct.

2. I agree, my inital thoughts were that they were the common case so it should be easy to use.

Reply:
I fixed `1` and added methods to log data vectors.